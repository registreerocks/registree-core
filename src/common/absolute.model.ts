import { ObjectType, Field, Int } from '@nestjs/graphql';

@ObjectType()
export class Absolute {
  @Field(_type => Int)
  absolute!: number;
}
