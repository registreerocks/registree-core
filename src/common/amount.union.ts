import { createUnionType } from '@nestjs/graphql';
import { Absolute } from './absolute.model';
import { Percentage } from './percentage.model';
import { Average } from './average.model';

export const AmountUnion = createUnionType({
  name: 'Amount',
  types: () => [Absolute, Percentage, Average],
  resolveType(value) {
    // XXX: Return string references instead of type instances, to work around upstream
    //      @nestjs/graphql bug.
    // See: https://github.com/registreerocks/registree-core/issues/367
    if ((value as Absolute).absolute) {
      return 'Absolute';
    }
    if ((value as Percentage).percentage) {
      return 'Percentage';
    }
    if ((value as Average).average) {
      return 'Average';
    }
    return null;
  },
});

export const getUnionValue = (amount: Percentage | Absolute | Average) => {
  switch (amount.amountType) {
    case 'Percentage':
      return amount.percentage;
    case 'Absolute':
      return amount.absolute;
    case 'Average':
      return amount.average;
  }
};

export const getDegreeAmountAsUnion = (degree: {
  absolute?: number;
  percentage?: number;
  average?: number;
}): Absolute | Percentage | Average => {
  if (degree.absolute) {
    return { absolute: degree.absolute, amountType: 'Absolute' };
  } else if (degree.average) {
    return { average: degree.average, amountType: 'Average' };
  } else if (degree.percentage) {
    return { percentage: degree.percentage, amountType: 'Percentage' };
  } else {
    return { percentage: 0, amountType: 'Percentage' };
  }
};
