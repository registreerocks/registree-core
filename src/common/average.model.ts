import { ObjectType, Field, Int } from '@nestjs/graphql';

@ObjectType()
export class Average {
  @Field(_type => Int)
  average!: number;

  readonly amountType: 'Average' = 'Average';
}
