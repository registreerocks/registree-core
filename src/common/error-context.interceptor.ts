import {
  Injectable,
  NestInterceptor,
  ExecutionContext,
  CallHandler,
  HttpException,
  ContextType,
} from '@nestjs/common';
import { Observable, MonoTypeOperatorFunction, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { ApolloError } from 'apollo-server-core';
import { WrappedError } from './errors/wrapped.error';

@Injectable()
export class ErrorContextInterceptor implements NestInterceptor {
  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
    const meta = {
      handlerClass: context.getClass().name,
      handler: context.getHandler().name,
      contextType: context.getType(),
    };
    return next.handle().pipe(this.rxjsErrorHandler(meta));
  }

  rxjsErrorHandler({
    handlerClass,
    handler,
    contextType,
  }: {
    handlerClass: string;
    handler: string;
    contextType: ContextType;
  }): MonoTypeOperatorFunction<void> {
    return catchError(err => {
      if (err instanceof ApolloError) {
        return throwError(err);
      } else if (err instanceof HttpException && contextType === 'http') {
        return throwError(err);
      } else {
        return throwError(new WrappedError(err, handlerClass, handler));
      }
    });
  }
}
