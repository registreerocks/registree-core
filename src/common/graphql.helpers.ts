import { User } from './interfaces/user.interface';

/**
 * Check and get the user from the GraphQL resolver context.
 *
 * @throw {Error} If the context has no user.
 */
export function userFromGqlContext(ctx: { req?: { user?: User } }): User {
  if (!ctx.req) {
    throw new Error(
      'userFromGqlContext: request missing from resolver context',
    );
  } else if (!ctx.req.user) {
    throw new Error(
      'userFromGqlContext: user missing from context request (Did authentication middleware run?)',
    );
  } else {
    return ctx.req.user;
  }
}
