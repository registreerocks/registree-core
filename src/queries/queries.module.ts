import { Module } from '@nestjs/common';
import { QueriesResolver } from './queries.resolver';
import { QueriesService } from './queries.service';
import { UploadModule } from 'src/upload/upload.module';

@Module({
  imports: [UploadModule],
  providers: [QueriesResolver, QueriesService],
})
export class QueriesModule {}
