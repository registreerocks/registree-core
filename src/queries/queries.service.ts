import { Injectable } from '@nestjs/common';
import { EventQuery } from './models/event-query.model';
import { CreateEventQueryInput } from './dto/create-event-query.input';
import { FileUpload } from 'graphql-upload';
import { QueryDataService } from 'src/query-data/query-data.service';
import { UploadService } from 'src/upload/upload.service';
import format from 'date-fns/format';
import { appConstants } from '../constants';
import { Quote } from 'src/pricing/models/quote.model';
import { PricingService } from 'src/pricing/pricing.service';
import { CreateQueryRequest } from 'src/query-data/dto/create-query.request';
import { AttachmentDto } from 'src/query-data/dto/attachment.dto';
import { mapEventQuery } from './mappers/map-event-query';
import { orderBy } from 'lodash';
import { UpdateEventInfoInput } from './dto/update-event-info.input';
import { UpdateEventRequest } from 'src/query-data/dto/update-event.request';

@Injectable()
export class QueriesService {
  constructor(
    private readonly uploadService: UploadService,
    private readonly queryDataService: QueryDataService,
    private readonly pricingService: PricingService,
  ) {}

  async getCustomerQueries(customerId: string): Promise<EventQuery[]> {
    const response = await this.queryDataService.getCustomerQueries(customerId);
    const mappedResponse = response.map(mapEventQuery);
    return orderBy(mappedResponse, [r => r.eventDetails.startDate], ['desc']);
  }

  async getQuote(input: CreateEventQueryInput): Promise<Quote> {
    const studentCount = await this.queryDataService.getStudentCountForQuery(
      this.createQueryRequestMapper(input, 'dry-run'),
    );
    return this.pricingService.getQuote(studentCount);
  }

  async getQuery(queryId: string): Promise<EventQuery> {
    const response = await this.queryDataService.getQuery(queryId);
    return mapEventQuery(response);
  }

  async createEventQuery(
    input: CreateEventQueryInput,
    customerId: string,
  ): Promise<EventQuery> {
    const attachments = await this.handleAttachments(input.attachments);

    const queryId = await this.queryDataService.createQuery(
      this.createQueryRequestMapper(input, customerId, attachments),
    );

    const response = await this.queryDataService.getQuery(queryId);
    return mapEventQuery(response);
  }

  async updateEventInfo(
    queryId: string,
    input: UpdateEventInfoInput,
  ): Promise<EventQuery> {
    const response = await this.queryDataService.updateEventInfo(
      queryId,
      this.updateEventRequestMapper(input),
    );
    return mapEventQuery(response);
  }

  async addAttachments(
    queryId: string,
    input: Promise<FileUpload>[],
  ): Promise<EventQuery> {
    const processedAttachments = await this.handleAttachments(input);
    const response = await this.queryDataService.addAttachments(
      queryId,
      processedAttachments,
    );
    return mapEventQuery(response);
  }

  async deleteAttachments(
    queryId: string,
    input: string[],
  ): Promise<EventQuery> {
    const response = await this.queryDataService.deleteAttachments(
      queryId,
      input,
    );
    return mapEventQuery(response);
  }

  private createQueryRequestMapper = (
    input: CreateEventQueryInput,
    customerId: string,
    attachments: { filename: string; id: string; mimetype: string }[] = [],
  ): CreateQueryRequest => ({
    customer_id: customerId,
    event: {
      address: input.address,
      end_date: format(input.endDate, appConstants.dateFormat),
      info: input.info,
      message: input.message,
      attachments,
      name: input.name,
      start_date: format(input.startDate, appConstants.dateFormat),
      type: input.eventType,
      password: input.password,
    },
    query: {
      details: input.degrees.map(d => ({
        degree_id: d.degreeId,
        absolute: d.absolute,
        percentage: d.percentage,
        degree_name: d.degreeName,
      })),
    },
  });

  private updateEventRequestMapper = (
    input: UpdateEventInfoInput,
  ): UpdateEventRequest => ({
    address: input.address,
    end_date: input.endDate
      ? format(input.endDate, appConstants.dateFormat)
      : undefined,
    info: input.info,
    message: input.message,
    name: input.name,
    password: input.password,
    start_date: input.startDate
      ? format(input.startDate, appConstants.dateFormat)
      : undefined,
    type: input.eventType,
  });

  private async handleAttachments(
    files: Promise<FileUpload>[] = [],
  ): Promise<AttachmentDto[]> {
    return Promise.all(
      files.map(async filePromise => {
        const file = await filePromise;
        if (appConstants.acceptedMimetypes.includes(file.mimetype)) {
          try {
            const key = await this.uploadService.saveFile(file);
            return {
              filename: file.filename,
              id: key,
              mimetype: file.mimetype,
            };
          } catch (error) {
            /* eslint-disable @typescript-eslint/no-unsafe-member-access */
            if (error.name === 'PayloadTooLargeError') {
              throw new Error(
                `The file size exceeds ${appConstants.fileSize} mb`,
              );
            } else {
              throw error;
            }
            /* eslint-enable @typescript-eslint/no-unsafe-member-access */
          }
        } else {
          throw new Error(
            `The following mimetype is not accepted: ${file.mimetype}`,
          );
        }
      }),
    );
  }
}
