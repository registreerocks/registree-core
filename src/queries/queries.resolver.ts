import { Resolver, Mutation, Args, Query, ID } from '@nestjs/graphql';
import { EventQuery } from './models/event-query.model';
import { CreateEventQueryInput } from './dto/create-event-query.input';
import { QueriesService } from './queries.service';
import { CurrentUser } from 'src/auth/current-user.decorator';
import { UseGuards } from '@nestjs/common';
import { GqlAuthGuard } from 'src/auth/gql-auth.guard';
import { findManyCursor } from 'src/common/pagination/find-many-cursors';
import { PaginationArgs } from 'src/common/pagination/pagination-args';
import { EventQueryConnection } from './models/pagination/event-query-connection.model';
import { paginateArray } from 'src/common/pagination/paginate-array';
import { User } from 'src/common/interfaces/user.interface';

@Resolver(_of => EventQuery)
export class QueriesResolver {
  constructor(private readonly queriesService: QueriesService) {}

  @Mutation(_returns => EventQuery)
  @UseGuards(GqlAuthGuard)
  async createQuery(
    @Args({ name: 'createEventQueryInput', type: () => CreateEventQueryInput })
    input: CreateEventQueryInput,
    @CurrentUser()
    user: User,
  ): Promise<EventQuery> {
    return this.queriesService.createEventQuery(input, user.dbId);
  }

  @Query(_returns => EventQueryConnection)
  async getQueries(
    @Args() { after, before, first, last }: PaginationArgs,
    @Args({ name: 'customerId', type: () => ID }) customerId: string,
  ): Promise<EventQueryConnection> {
    const queries = await this.queriesService.getCustomerQueries(customerId);
    // TODO: Default sorting
    const paginatedQueries = await findManyCursor(
      args => {
        return Promise.resolve(paginateArray(queries, args));
      },
      _args => {
        return Promise.resolve(queries.length);
      },
      { first, last, before, after },
    );
    return paginatedQueries;
  }
}
