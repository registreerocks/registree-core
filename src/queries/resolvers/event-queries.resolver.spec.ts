import { INestApplication } from '@nestjs/common';
import { GraphQLModule, GraphQLSchemaHost } from '@nestjs/graphql';
import { Test, TestingModule } from '@nestjs/testing';
import { gql } from 'apollo-server-core';
import fc from 'fast-check';
import { GraphQLSchema, printSchema } from 'graphql';
import {
  coerceNullPrototypeObjects,
  execGraphQL,
} from '../../common/test.helpers';
import { CustomersService } from '../../customers/customers.service';
import { Quote } from '../../pricing/models/quote.model';
import { RsvpCost } from '../../pricing/models/rsvp-cost.model';
import { PricingService } from '../../pricing/pricing.service';
import { arbitraryCreateEventQueryInput } from '../dto/create-event-query.input.arb';
import { QueriesService } from '../queries.service';
import { EventQueriesResolver } from './event-queries.resolver';

describe('EventQueriesResolver', () => {
  let app: INestApplication;
  let schema: GraphQLSchema;
  const mockQueriesService: Partial<QueriesService> = {};

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [GraphQLModule.forRoot({ autoSchemaFile: true })],
      providers: [
        { provide: QueriesService, useValue: mockQueriesService },
        { provide: PricingService, useValue: null },
        { provide: CustomersService, useValue: null },
        EventQueriesResolver,
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();

    const schemaHost: GraphQLSchemaHost = app.get(GraphQLSchemaHost);
    schema = schemaHost.schema;
  });

  afterAll(async () => {
    await app.close();
  });

  afterEach(() => {
    expect(mockQueriesService).toStrictEqual({});
  });

  test('getQuote', async () => {
    const query = gql`
      query getQuote($input: CreateEventQueryInput!) {
        quote: getQuote(createEventQueryInput: $input) {
          numberOfStudents
          rsvpCostBreakdown {
            cost
            percent
          }
        }
      }
    `;

    await fc.assert(
      fc.asyncProperty(
        arbitraryCreateEventQueryInput(),
        arbitraryQuote(),
        async (input, output) => {
          try {
            mockQueriesService.getQuote = jest
              .fn()
              .mockResolvedValueOnce(output);

            const result = await execGraphQL(schema, query, {
              variableValues: { input },
            });
            expect(coerceNullPrototypeObjects(result)).toStrictEqual({
              data: { quote: output },
            });
            // Verify calls:
            expect(mockQueriesService.getQuote).toHaveBeenCalledWith(input);
            expect(mockQueriesService.getQuote).toHaveBeenCalledTimes(1);
          } finally {
            delete mockQueriesService.getQuote;
          }
        },
      ),
    );
  });

  test('schema snapshot', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });
});

function arbitraryRsvpCost(): fc.Arbitrary<RsvpCost> {
  return fc.record({
    cost: fc.double(),
    percent: fc.integer(),
  });
}

function arbitraryQuote(): fc.Arbitrary<Quote> {
  return fc.record({
    numberOfStudents: fc.integer(),
    rsvpCostBreakdown: fc.array(arbitraryRsvpCost()),
  });
}
