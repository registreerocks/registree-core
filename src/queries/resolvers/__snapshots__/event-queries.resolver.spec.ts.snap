// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventQueriesResolver schema snapshot 1`] = `
"type BillingDetails {
  addressee: String
  city: String
  country: String
  email: String
  line1: String
  line2: String
  province: String
  vat: String
  zip: String
}

type Contact {
  email: String
  name: String
  userId: ID!
  dbId: ID!
  calendlyLink: String
}

type Customer {
  id: ID!
  description: String
  name: String!
  billingDetails: BillingDetails!
  contacts: [Contact!]!
}

type Degree {
  id: ID!
  name: String!
  description: String!
  faculty: Faculty!
  level: String!
}

type GroupedDegrees {
  name: String!
  degrees: [Degree!]!
}

type Faculty {
  id: ID!
  name: String!
  description: String!
  university: University!
  groupedDegrees: [GroupedDegrees!]!
  degrees: [Degree!]!
}

type University {
  id: ID!
  country: String!
  name: String!
  physicalAddress: String!
  shortName: String!
  faculties: [Faculty!]!
}

type UploadedFile {
  id: ID!
  filename: String!
  mimetype: String!
  url: String!
}

type Invitation {
  accepted: Boolean!
  attended: Boolean!
  respondedAt: DateTime
  sentAt: DateTime
  viewedAt: DateTime
  email: String
  invitationState: InvitationState!
}

\\"\\"\\"
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
\\"\\"\\"
scalar DateTime

enum InvitationState {
  NONE
  VIEWED
  RESPONDED
  ACCEPTED
  ATTENDED
}

type EventMetrics {
  acceptedCount: Int!
  attendedCount: Int!
  respondedCount: Int!
  viewedCount: Int!
}

type EventDetails {
  address: String!
  startDate: DateTime!
  endDate: DateTime!
  attachments: [UploadedFile!]!
  information: String!
  message: String
  name: String!
  eventType: String!
  invites: [Invitation!]!
  metrics: EventMetrics!
  password: String
}

type DegreeSelection {
  amount: Amount!
  degree: Degree!
}

union Amount = Absolute | Percentage | Average

type Absolute {
  absolute: Int!
}

type Percentage {
  percentage: Int!
}

type Average {
  average: Int!
}

type PageInfo {
  endCursor: String
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
}

type Student {
  email: String!
  name: String!
  userId: ID!
  studentNumber: ID!
}

type StudentLink {
  id: ID!
  student: Student
}

type QueryTranscript {
  id: ID!
  degreeAverage: Float!
  degreeCompleted: Boolean!
  latestTerm: Float!
  degree: Degree!
  studentLink: StudentLink!
}

type QueryTranscriptEdge {
  cursor: String!
  node: QueryTranscript!
}

type QueryTranscriptConnection {
  edges: [QueryTranscriptEdge!]
  pageInfo: PageInfo!
  totalCount: Int!
}

type QueryDetails {
  parameters: [DegreeSelection!]!
  academicYearOfStudyList: [AcademicYearOfStudy!]!
  faculties: [Faculty!]!
  results: [QueryTranscriptConnection!]!
  updatedAt: DateTime!
  race: [String!]!
  gender: [String!]!
  smsMessage: String!
}

\\"\\"\\"The academic year of study a student is in.\\"\\"\\"
enum AcademicYearOfStudy {
  YEAR_1
  YEAR_2
  YEAR_3_PLUS
}

type RsvpCost {
  cost: Float!
  percent: Int!
}

type Quote {
  numberOfStudents: Int!
  rsvpCostBreakdown: [RsvpCost!]!
}

type EventQuery {
  id: ID!
  customer: Customer!
  eventDetails: EventDetails!
  queryDetails: QueryDetails!
  quoteDetails: Quote!
  currentPrice: Float!
}

type EventQueryEdge {
  cursor: String!
  node: EventQuery!
}

type EventQueryConnection {
  edges: [EventQueryEdge!]
  pageInfo: PageInfo!
  totalCount: Int!
}

type EventSummary {
  eventId: ID!
  eventName: String!
  invitedCount: Int!
  acceptedCount: Int!
  attendedCount: Int!
  respondedCount: Int!
  viewedCount: Int!
}

type Query {
  getQuote(createEventQueryInput: CreateEventQueryInput!): Quote!
  getQueries(customerId: ID!, after: String, before: String, first: Int, last: Int): EventQueryConnection!
  getEventSummaries(customerId: ID!): [EventSummary!]!
  getStudentQueries(filter: StudentQueryFilter, after: String, before: String, first: Int, last: Int): EventQueryConnection!
  getQuery(id: ID!): EventQuery!
}

input CreateEventQueryInput {
  name: String!
  address: String!
  startDate: DateTime!
  endDate: DateTime!
  information: String!
  message: String
  degrees: [DegreeInput!]!
  academicYearOfStudyList: [AcademicYearOfStudy!]!
  race: [String!]
  gender: [String!]
  smsMessage: String
  eventType: String!
  attachments: [Upload!]
  password: String
}

input DegreeInput {
  degreeId: String!
  absolute: Int
  percentage: Int
  average: Int
}

\\"\\"\\"The \`Upload\` scalar type represents a file upload.\\"\\"\\"
scalar Upload

input StudentQueryFilter {
  invitationState: InvitationState
}

type Mutation {
  createQuery(createEventQueryInput: CreateEventQueryInput!): EventQuery!
  expandQuery(expandEventQueryInput: ExpandEventQueryInput!, queryId: ID!): EventQuery!
}

input ExpandEventQueryInput {
  degrees: [DegreeInput!]!
  academicYearOfStudyList: [AcademicYearOfStudy!]!
}
"
`;
